services/agit/agit.go

```diff
@@ -0,0 +1,288 @@
+				BaseRepo:     repo,
+				MergeBase:    "",
+				Type:         models.PullRequestGitea,
+				Flow:         models.PullRequestFlowAGit,
+			}
+
+			if err := pull_service.NewPullRequest(repo, prIssue, []int64{}, []string{}, pr, []int64{}); err != nil {
+				if models.IsErrUserDoesNotHaveAccessToRepo(err) {
+					ctx.Error(http.StatusBadRequest, "UserDoesNotHaveAccessToRepo", err.Error())
+					return nil
+				}
+				ctx.Error(http.StatusInternalServerError, "NewPullRequest", err.Error())
+				return nil
+			}
+
+			log.Trace("Pull request created: %d/%d", repo.ID, prIssue.ID)
+
+			results = append(results, private.HockProcReceiveRefResult{
+				Ref:         pr.GetGitRefName(),
+				OriginalRef: opts.RefFullNames[i],
+				OldOID:      git.EmptySHA,
+				NewOID:      opts.NewCommitIDs[i],
+			})
+			continue
+		}
+
+		// update exist pull request
+		if err := pr.LoadBaseRepo(); err != nil {
+			log.Error("Unable to load base repository for PR[%d] Error: %v", pr.ID, err)
+			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+				"Err": fmt.Sprintf("Unable to load base repository for PR[%d] Error: %v", pr.ID, err),
+			})
+			return nil
+		}
+
+		oldCommitID, err := gitRepo.GetRefCommitID(pr.GetGitRefName())
+		if err != nil {
+			log.Error("Unable to get ref commit id in base repository for PR[%d] Error: %v", pr.ID, err)
+			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+				"Err": fmt.Sprintf("Unable to get ref commit id in base repository for PR[%d] Error: %v", pr.ID, err),
+			})
+			return nil
+		}
+
+		if oldCommitID == opts.NewCommitIDs[i] {
+			results = append(results, private.HockProcReceiveRefResult{
+				OriginalRef: opts.RefFullNames[i],
+				OldOID:      opts.OldCommitIDs[i],
+				NewOID:      opts.NewCommitIDs[i],
+				Err:         "new commit is same with old commit",
+			})
+			continue
+		}
+
+		if !forcePush {
+			output, err := git.NewCommand("rev-list", "--max-count=1", oldCommitID, "^"+opts.NewCommitIDs[i]).RunInDirWithEnv(repo.RepoPath(), os.Environ())
+			if err != nil {
+				log.Error("Unable to detect force push between: %s and %s in %-v Error: %v", oldCommitID, opts.NewCommitIDs[i], repo, err)
+				ctx.JSON(http.StatusInternalServerError, private.Response{
+					Err: fmt.Sprintf("Fail to detect force push: %v", err),
+				})
+				return nil
+			} else if len(output) > 0 {
+				results = append(results, private.HockProcReceiveRefResult{
+					OriginalRef: oldCommitID,
+					OldOID:      opts.OldCommitIDs[i],
+					NewOID:      opts.NewCommitIDs[i],
+					Err:         "request `force-push` push option",
+				})
+				continue
+			}
+		}
+
+		pr.HeadCommitID = opts.NewCommitIDs[i]
+		if err = pull_service.UpdateRef(pr); err != nil {
+			log.Error("Failed to update pull ref. Error: %v", err)
+			ctx.JSON(http.StatusInternalServerError, map[string]interface{}{
+				"Err": fmt.Sprintf("Failed to update pull ref. Error: %v", err),
+			})
+			return nil
+		}
+
+		pull_service.AddToTaskQueue(pr)
```

Comment:
@a1012112796 Do we need to call `pull_service.AddToTaskQueue` (`pull_service.StartPullRequestCheckImmediately` in new code) here?


The PR check would also be added in the post-receive hook?